Modules/itertoolsmodule.c

```diff
@@ -330,21 +330,27 @@ pairwise_next(pairwiseobject *po)
         return NULL;
     }
     if (old == NULL) {
-        po->old = old = (*Py_TYPE(it)->tp_iternext)(it);
+        old = (*Py_TYPE(it)->tp_iternext)(it);
         if (old == NULL) {
             Py_CLEAR(po->it);
+            Py_CLEAR(po->old);
             return NULL;
         }
     }
+    else {
+        Py_INCREF(old);
```

Comment:
Other simple solution is setting `po->old = NULL` instead of increfing `old`, so `old` holds the only reference. It only differs from the present solution in case `check(3, ...)`, and I tested much more test cases than present in this PR, including multiple re-entering points.